### PR TITLE
GEOMESA-939 Fix Exponential Query Planning Time for High Cardinality …

### PR DESCRIPTION
**Pull request by jahhulbert-ccri**
…OR attribute queries

### CODE REVIEW

**Comment by elahrvivaz** [comment]
looks good to me

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/QueryFilterSplitter.scala [hunk 1]
```diff
@@ -223,7 +224,45 @@ class QueryFilterSplitter(sft: SimpleFeatureType) extends Logging {
       }
     }
 
-    makeDisjoint(mergeOverlappedFilters(combineSecondaryFilters(reduceChildOptions(getChildOptions))))
+    // detect the pattern it the pattern (attr in (1,2,3,4,5,6, etc) AND something else)
```
Comment (anthonyccri) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/QueryFilterSplitter.scala]: [nitpick] comment doesn't make sense

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/QueryFilterSplitter.scala [hunk 2]
```diff
@@ -223,7 +224,45 @@ class QueryFilterSplitter(sft: SimpleFeatureType) extends Logging {
...
+    def detectSingleAttrOr: Boolean = {
+      // each child is expressed as (attr = x(i) AND expr2 AND expr3 ...)
+      val attrs =
+        filter.getChildren.map {
+          case and: And => and.getChildren.flatMap {
```
Comment (anthonyccri) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/QueryFilterSplitter.scala]: Aren't we dealing with ```OR```s here, not ```AND```s.  Shouldn't it be something like:

```sql
bbox(geom, ...) AND dtg between .... AND (attr = '1' OR attr = '2' ....)
```

Comment (jahhulbert-ccri) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/QueryFilterSplitter.scala]: We are...but the ```logicDistributionDNF``` method transforms it out into something like this:
```
[
  [
     [ high = h0 ]
     AND [ geom bbox POLYGON ((40 40, 40 50, 50 50, 50 40, 40 40)) ] 
     AND org.geotools.filter.temporal.DuringImpl@e5a72d93
  ] 
  OR 
  [
    [ high = h1 ]
    AND [ geom bbox POLYGON ((40 40, 40 50, 50 50, 50 40, 40 40)) ]
    AND org.geotools.filter.temporal.DuringImpl@e5a72d93
  ]
  OR 
  [
    [ high = h2 ]
```

So really the  match statement is referring to the "children" of the OR being an AND: 
```
[[ high = h0 ] AND [ geom bbox POLYGON ((40 40, 40 50, 50 50, 50 40, 40 40)) ] AND org.geotools.filter.temporal.DuringImpl@e5a72d93]
```

That being said, you bring up a good point that we should verify it is a top level OR filter whose children are  AND predicates.

Comment (jahhulbert-ccri) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/QueryFilterSplitter.scala]: also note that i think we may want to redesign some of the code anyways to make it easier to follow/inject new optimizations :)

File: geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/QueryFilterSplitter.scala (hunk 1 above)
Comment (jahhulbert-ccri) [inline geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/index/QueryFilterSplitter.scala]: added better comment

**Comment by jahhulbert-ccri** [comment]
@anthonyccri @elahrvivaz added an update...lemme know if there's anything else...it's a hairy piece area of the codebase

**Comment by elahrvivaz** [comment]
yeah, trying to figure out OR combinations was tricky - if there's a better way, I'm all for refactoring this code.

**Comment by jahhulbert-ccri** [comment]
Yeah, my gut feels there are some better ways, but they will take much more thought and engineering...that's what I want GEOMESA-941 to be...this was the quick fix for operational purposes :) We'll need to think more about a clean way to do it since its difficult.

**Comment by anthonyccri** [comment]
e976c7c..71aec6d

**Pull request closed without merging**